crates/count-min-sketch/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+    /// sketch.
+    pub fn estimate<T: Hash>(&self, item: T) -> i64 {
+        let buckets = self
+            .hashfuncs
+            .iter()
+            .map(|h| h.hash_into_buckets(&item, self.width));
+
+        self.counters
+            .iter()
+            .zip(buckets)
+            .map(|(counter, bucket)| counter[bucket])
+            .min()
+            .unwrap()
+    }
+
+    /// Returns a vector of the indices for the buckets into which `item` hashes.
+    ///
+    /// The vector will have `self.depth` elements, each in the range
+    /// [0, self.width-1].
+    pub fn get_bucket_indices<T: Hash>(&self, item: T) -> Vec<usize> {
+        self.hashfuncs
+            .iter()
+            .map(|h| h.hash_into_buckets(&item, self.width))
+            .collect()
+    }
+
+    /// Adds the given `item` to the sketch.
+    pub fn add_value<T: Hash>(&mut self, item: T) {
+        for i in 0..self.depth {
+            let bucket = self.hashfuncs[i].hash_into_buckets(&item, self.width);
+            self.counters[i][bucket] += 1;
+        }
+    }
+
+    /// Subtract the given `item` from the sketch.
+    pub fn subtract_value<T: Hash>(&mut self, item: T) {
+        for i in 0..self.depth {
+            let bucket = self.hashfuncs[i].hash_into_buckets(&item, self.width);
+            self.counters[i][bucket] -= 1;
+        }
+    }
+
+    /// Includes the counts from `other` into `self` via elementwise addition of
+    /// the counter vectors.
+    ///
+    /// The underlying `CountMinHashFn`s in each sketch must have the same keys.
+    pub fn combine(&mut self, other: CountMinSketch) {
+        assert_eq!(self.width, other.width);
+        assert_eq!(self.depth, other.depth);
+        assert_eq!(self.hashfuncs, other.hashfuncs);
+        for (counter1, counter2) in self.counters.iter_mut().zip(other.counters) {
+            for (val1, val2) in counter1.iter_mut().zip(counter2) {
+                *val1 += val2;
+            }
+        }
+    }
+}
+
+impl fmt::Display for CountMinSketch {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "Count-Min Sketch:")?;
+        write!(f, "+------++")?;
+        for _ in 0..self.width {
+            write!(f, "--------+")?;
+        }
+        writeln!(f)?;
+
+        write!(f, "|      ||")?;
+        for b in 0..self.width {
+            write!(f, "    {:>3} |", b)?;
+        }
+        writeln!(f)?;
+
+        write!(f, "+======++")?;
+        for _ in 0..self.width {
+            write!(f, "========+")?;
+        }
+        writeln!(f)?;
+
+        for n in 0..self.depth {
+            write!(f, "|  {:>3} ||", n)?;
+            for x in &self.counters[n] {
+                write!(f, " {:>6} |", x)?;
+            }
+            writeln!(f)?;
+        }
+
+        write!(f, "+------++")?;
+        for _ in 0..self.width {
+            write!(f, "--------+")?;
+        }
+        writeln!(f)
+    }
+}
+
+#[cfg(test)]
```

Comment:
Looks like these are all purely in terms of public interface (good) and therefore can move to `crates/count-min-sketch/tests/lib.rs`